### ISSUE TITLE
Overly verbose generated sealed trait ADT schema

### ISSUE THREAD

**Issue by harpocrates** [julienrf/endpoints#503]
When generating JSON schemas for simple sealed trait ADTs, there is often redundant information in the child schemas. In particular, the children reference the parent schema, which then indirectly references the other ADT variants. This leads to confusing schemas for otherwise simple types.

----

Consider the following simple ADT called `Parent`, an endpoint definition, and the generated OpenAPI docs:

```scala
import endpoints.openapi.model.OpenApi
import endpoints.{algebra, generic, openapi, Invalid, Valid}
import endpoints.generic.{docs, name}

sealed trait Parent
object Parent {
  case object Child1 extends Parent
  case class Child2(foo: Int) extends Parent
}

trait Alg
  extends algebra.Endpoints
  with algebra.JsonEntitiesFromSchemas
  with generic.JsonSchemas {

  implicit val parentFormat = genericJsonSchema[Parent]

  val parentGet = endpoint(get(path / "parent"), response = ok(jsonResponse[Parent]))
}

object AlgDocs
  extends Alg
  with openapi.Endpoints
  with openapi.JsonEntitiesFromSchemas {

  val api: openapi.model.OpenApi =
    openApi(openapi.model.Info("test", "0"))(parentGet)

  val jsonApi: String = OpenApi.stringEncoder.encode(AlgDocs.api)
}
```

this generates the following OpenAPI JSON:

<details>

```json
{
  "openapi": "3.0.0",
  "info": {
    "title": "test",
    "version": "0"
  },
  "paths": {
    "/parent": {
      "get": {
        "responses": {
          "400": {
            "description": "Client error",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/endpoints.Errors"
                }
              }
            }
          },
          "500": {
            "description": "Server error",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/endpoints.Errors"
                }
              }
            }
          },
          "200": {
            "description": "",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Parent"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "endpoints.Errors": {
        "type": "array",
        "items": {
          "type": "string"
        }
      },
      "Parent": {
        "oneOf": [
          {
            "$ref": "#/components/schemas/Parent.Child1"
          },
          {
            "$ref": "#/components/schemas/Parent.Child2"
          }
        ],
        "discriminator": {
          "propertyName": "type",
          "mapping": {
            "Child1": "#/components/schemas/Parent.Child1",
            "Child2": "#/components/schemas/Parent.Child2"
          }
        }
      },
      "Parent.Child1": {
        "allOf": [
          {
            "$ref": "#/components/schemas/Parent"
          },
          {
            "type": "object",
            "properties": {
              "type": {
                "type": "string",
                "enum": [
                  "Child1"
                ],
                "example": "Child1"
              }
            },
            "required": [
              "type"
            ]
          }
        ]
      },
      "Parent.Child2": {
        "allOf": [
          {
            "$ref": "#/components/schemas/Parent"
          },
          {
            "type": "object",
            "properties": {
              "type": {
                "type": "string",
                "enum": [
                  "Child2"
                ],
                "example": "Child2"
              },
              "foo": {
                "type": "integer",
                "format": "int32"
              }
            },
            "required": [
              "type",
              "foo"
            ]
          }
        ]
      }
    },
    "securitySchemes": {}
  }
}
```

</details>

...which turns into this in Swagger UI:

<img width="458" alt="Screen Shot 2020-03-23 at 9 04 47 PM" src="https://user-images.githubusercontent.com/10766081/77377625-09582300-6d4a-11ea-8b9e-9248fecaa83c.png">

I find the `oneOf` sections of `Parent.Child1` and `Parent.Child2` surprising since they add no additional information to the schema!

**Comment by harpocrates** [comment]
I see that this is the cause for including the parent schema in coproduct alternatives

https://github.com/julienrf/endpoints/blob/58a2cc7b92fe375f6400477d58f6dbbf650cb36e/openapi/openapi/src/main/scala/endpoints/openapi/JsonSchemas.scala#L483-L495

Looking through `git blame`, it looks as though this change is intentional and from https://github.com/julienrf/endpoints/pull/135 (this comment summarizes the possible encodings and the justification behind the current choice: https://github.com/julienrf/endpoints/pull/135#issuecomment-428164930). It would still be nice to have this be configurable though.

**Comment by julienrf** [comment]
I agree that it would be nice to have it at least configurable. Also, we lack of objectivity on how well do OpenAPI code generators work with the OpenAPI documents generated by _endpoints_.

**Comment by harpocrates** [comment]
The following are two places I think we could thread through information to configure this:

  * add the additional information about desired coproduct schema encoding into the combinators that produce `Tagged` schemas in `algebra.JsonSchemas`
  * add some overrideable `def coproductEncoding: CoproductEncoding` to `openapi.JsonSchemas` which acts as a global switch

I personally lean towards the second alternative. The choice of coproduct encoding seems to be motivated by which downstream tools are expected to use the OpenAPI spec, so it makes sense to have the option applied to all coproduct schemas at once.

As for alternatives, I took another look at https://github.com/julienrf/endpoints/pull/135#issuecomment-428164930, and it looks like only options a) and c) really make sense for coproduct encoding in _endpoints_. The issue with b) is that children type schemas won't be included in the final generated OpenAPI JSON since there aren't any references to them.

Thoughts? I've got a candidate implementation here: https://github.com/harpocrates/endpoints/commit/a14fe6eb823249b13a3e6447577f6d10036efb94.
